Set CORS preflight requests' `mode` to `cors`

ghostd opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #23780 (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

ghostd committed:
Set CORS preflight requests' `mode` to `cors`

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 5ad47ec has been approved by `jdm`

<!-- @bors-servo r=jdm 5ad47ec8d9fc878cc8f52ef082deea952f6c9a64 -->

bors-servo commented:
:hourglass: Testing commit 5ad47ec8d9fc878cc8f52ef082deea952f6c9a64 with merge 285d5095aab148ecddafb7015e33093d54cb9bd2...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/UQtAgMPaQ6q8eEGeNn0DbQ)

ghostd commented:
I'll update the test results this evening or tomorrow.

ghostd commented:
I rebased and updated the "PASS [expected FAIL]" tests. I still have a "FAIL [expected PASS]" with fetch/api/cors/cors-preflight-redirect.any.js

If i remove the "Accept: */*" insertion (the new "Step 2" of the algorithm), the test pass (actually with the same results as before), but with the Accept header the test fails... i'm still digging...

ghostd commented:
I updated more implementation than specifying by the issue. I still have some "regressions" with the tests from `cors-preflight-redirect.any.js`. I still don't know why they pass on master (or what part of `fetch` spectification indicates the redirections should fail... if someone can point me the right section...) Maybe i still need to update more code.

I really need someone review the new `step 4` and `step 7` in the `cors_preflight_fetch` function against the specification, more precisely i'm not sure i correctly set the `()` in the conditions.

ghostd commented on tests/wpt/metadata/fetch/api/cors/cors-preflight-not-cors-safelisted.any.js.ini:
```diff
@@ -1,60 +1,18 @@
-    expected: FAIL
-
   [Need CORS-preflight for content-language/\x01 header]
     expected: FAIL
```
These tests still fail because of:
https://github.com/servo/servo/blob/a4bf7e51793813b2f5120c621a02771ad495be92/components/script/dom/headers.rs#L361-L363

Should we provide a PR to change HeaderValue?

jdm replied:
We should file an issue about it, at least.

ghostd replied:
Done #27764 27764

ghostd commented on tests/wpt/metadata/fetch/api/cors/cors-preflight.any.js.ini:
```diff
@@ -6,27 +6,6 @@
   [CORS [PUT\] [several headers\], server allows]
```
These tests fail because of https://github.com/servo/servo/blob/a4bf7e51793813b2f5120c621a02771ad495be92/components/script/dom/headers.rs#L88-L100

Should i remove the leading space?

I'm not sure to understand the point `HyperHeaders might not expose the information we need to do it right`...

jdm replied:
I feel like there's possibly some history here. File an issue about this, please!

ghostd replied:
Done #27765

ghostd commented:
So, if i had did a search into the open issues before, i saved many hours :'( [This issue](https://github.com/web-platform-tests/wpt/issues/20517) already reports what i previously called "regressions". With the current implementation we have the [same results than the other browsers](https://wpt.fyi/results/fetch/api/cors/cors-preflight-redirect.any.html?label=experimental&label=master&aligned).

Can someone review this please?

ghostd commented:
ping @jdm @CYBAI

bors-servo commented:
:umbrella: The latest upstream changes (presumably #27726) made this pull request unmergeable. Please resolve the merge conflicts.

jdm reviewed (approved):
This looks great! Apologies for the long delay in reviewing it.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 5b40068 has been approved by `jdm`

<!-- @bors-servo r=jdm 5b400685876636af796ded1e64517d274fb18d47 -->

bors-servo commented:
:hourglass: Testing commit 5b400685876636af796ded1e64517d274fb18d47 with merge f8ebe03dc287797834cdf87c78de8ff3a9525575...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/O0sVKC3PSluqkogQytDJ0A)
Approved by: jdm
Pushing f8ebe03dc287797834cdf87c78de8ff3a9525575 to master...
